progenitor-impl/src/lib.rs

```diff
@@ -304,6 +312,42 @@ impl Generator {
 
         let types = self.type_space.to_stream();
 
+        let extra_impl = TokenStream::from_iter(
+            self.forms
+                .iter()
+                .map(|type_id| {
+                    let typ = self.get_type_space().get_type(type_id).unwrap();
+                    let td = typ.details();
+                    let TypeDetails::Struct(tstru) = td else { unreachable!() };
+                    let properties = indexmap::IndexMap::<&'_ str, _>::from_iter(
+                        tstru
+                            .properties()
+                            .filter_map(|(prop_name, prop_id)| {
+                                self.get_type_space()
+                                    .get_type(&prop_id).ok()
+                                    .map(|prop_typ| (prop_name, prop_typ))
+                            })
+                        );
+                    let properties = syn::punctuated::Punctuated::<_, syn::Token![,]>::from_iter(
+                        properties
+                            .into_iter()
+                            .map(|(prop_name, _prop_ty)| {
+                                let ident = quote::format_ident!("{}", prop_name);
+                                quote!{ (#prop_name, &self. #ident) }
+                            }));
+
+                    let form_name = quote::format_ident!("{}",typ.name());
+
+                    quote! {
+                        impl #form_name {
+                            pub fn as_form<'f>(&'f self) -> impl std::iter::Iterator<Item=(&'static str, &'f [u8])> {
+                                [#properties]
+                                    .into_iter()
+                                    .filter_map(|(name, val)| val.as_ref().map(|val| (name, val.as_bytes())))
```

Comment:
When trying to compile I'm getting the following error (rust v1.78.0):
```
     --> <redacted>/src/lib.rs:12324:47
      |
12324 |                 .filter_map(|(name, val)| val.as_ref().map(|val| (name, val.as_bytes())))
      |                                               ^^^^^^   --- type must be known at this point
      |
help: try using a fully qualified path to specify the expected types
      |
12324 |                 .filter_map(|(name, val)| <std::string::String as AsRef<T>>::as_ref(val).map(|val| (name, val.as_bytes())))
      |                                           ++++++++++++++++++++++++++++++++++++++++++   ~
```

The specification I'm using includes a single property.

Reply:
I ran into this limitation a few weeks back, and the root cause here being unified way between `&str`, `Vec<u8>`, and others to retrieve a byte slice. Adding a custom trait and implementing fixes that.

That said, It'll be September before I'll be able to work on this PR